ballerina/channel_csv_io.bal

```diff
@@ -40,30 +40,90 @@ Error {
-    WritableCSVChannel csvChannel = check getWritableCSVChannel(writableChannel);
-    if content is string[][] {
-        foreach string[] r in content {
+isolated function channelWriteCsv(string path, FileWriteOption option, string[][]|map<anydata>[] contentToWrite) returns Error? {
+
+    WritableCSVChannel csvChannel;
+    if contentToWrite is string[][] {
+        csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        foreach string[] r in contentToWrite {
             Error? writeResult = csvChannel.write(r);
             if writeResult is Error {
                 check csvChannel.close();
                 return writeResult;
             }
         }
-    } else if content is map<anydata>[] {
-        foreach map<anydata> row in content {
+        check csvChannel.close();
+    } else if contentToWrite is map<anydata>[] {
+        string[] headers = [];
+        if contentToWrite.length() == 0 {
+            return;
+        }
+        if option == OVERWRITE {
+            headers = contentToWrite[0].keys();
+            csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+            Error? headerWriteResult = csvChannel.write(headers);
+            if headerWriteResult is Error {
+                check csvChannel.close();
+                return headerWriteResult;
+            }
+        } else {
+            headers = [];
+            stream<string[], Error?>|Error csvContent = fileReadCsvAsStream(path);
+            if (csvContent !is Error) {
+                var csvLine = csvContent.next();
+                check csvContent.close();
+                if csvLine !is error {
+                    if csvLine !is () {
+                        if csvLine["value"] != [""] {
+                            headers = csvLine["value"];
+                        }
+                    }
+                } else { 
+                    return csvLine;
+                }
+            } else if csvContent is FileNotFoundError {
+                headers = [];
+            } else {
+                return csvContent;
+            }
+            if headers.length() > 0 {
+                csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+                if contentToWrite[0].keys().length() != headers.length() {
+                    check csvChannel.close();
+                    GenericError e = error GenericError("CSV file and the Record structure do not  match.");
+                    return e;
+                }
+                foreach string header in headers {
+                    int|Error? key = contentToWrite[0].keys().lastIndexOf(header.trim());
+                    if key is Error || key is () {
+                        check csvChannel.close();
+                        GenericError e = error GenericError("CSV file and the Record structure do not  match.");
```

Comment:
give a more specific error message. we can give which header doesn't match here

Reply:
updated in [9967638](https://github.com/ballerina-platform/module-ballerina-io/pull/827/commits/9967638b794331da447a8be1d99326691ec4e69f)